rts/Rml/Backends/RmlUi_Backend.cpp

```diff
@@ -0,0 +1,495 @@
+#include "Rml/SolLua/RmlSolLua.h"
+#include "RmlUi_Backend.h"
+
+#ifndef HEADLESS
+#include "RmlUi_Renderer_GL3_Recoil.h"
+#else
+#include "RmlUi_Renderer_Headless.h"
+#endif
+
+#include "RmlUi_SystemInterface.h"
+#include "RmlUi_VFSFileInterface.h"
+#include "System/Input/InputHandler.h"
+#include "System/Log/ILog.h"
+
+using CtxMutex = std::recursive_mutex;
+using CtxLockGuard = std::lock_guard<CtxMutex>;
+struct lua_State;
+
+/// Passes through RML events to the function pointers given in the constructor
+class PassThroughPlugin : public Rml::Plugin
+{
+	void (*onContextCreate)(Rml::Context*);
+	void (*onContextDestroy)(Rml::Context*);
+
+public:
+	PassThroughPlugin(void (*onContextCreate)(Rml::Context*),
+	                  void (*onContextDestroy)(Rml::Context*))
+		: onContextCreate{onContextCreate}
+		, onContextDestroy{onContextDestroy}
+	{
+	}
+
+	int GetEventClasses() override
+	{
+		return EVT_BASIC;
+	}
+
+	void OnInitialise() override{};
+	void OnShutdown() override{};
+
+	void OnContextCreate(Rml::Context* context) override
+	{
+		onContextCreate(context);
+	};
+
+	void OnContextDestroy(Rml::Context* context) override
+	{
+		onContextDestroy(context);
+	};
+};
+
+/**
+  Global data used by this backend.
+  Lifetime governed by the calls to Backend::Initialize() and
+  Backend::Shutdown().
+ */
+struct BackendData {
+	RmlSystemInterface system_interface;
+#ifndef HEADLESS
+	RenderInterface_GL3_Recoil render_interface;
+#else
+	RenderInterface_Headless render_interface;
+#endif
+	VFSFileInterface file_interface;
+
+	SDL_Window* window = nullptr;
+	SDL_GLContext gl_context = nullptr;
+	std::vector<Rml::Context*> contexts;
+	InputHandler::SignalType::connection_type inputCon;
+	CRmlInputReceiver inputReceiver;
+
+	// make atomic_bool?
+	bool initialized = false;
+	bool debuggerAttached = false;
+	int winX = 1;
+	int winY = 1;
+
+	lua_State* ls = nullptr;
+	Rml::SolLua::SolLuaPlugin* luaPlugin = nullptr;
+
+	CtxMutex contextMutex;
+	Rml::UniquePtr<PassThroughPlugin> plugin;
+    Rml::UniquePtr<Rml::ElementInstancerGeneric<RmlGui::ElementLuaTexture>> element_lua_texture_instancer;
+};
+
+static Rml::UniquePtr<BackendData> data;
+
+bool RmlInitialized()
+{
+	return data && data->initialized;
+}
+
+bool RmlGui::Initialize(SDL_Window* target_window, SDL_GLContext target_glcontext, int winX,
+                        int winY)
+{
+	data = Rml::MakeUnique<BackendData>();
+
+	if (!data->render_interface) {
+		data.reset();
+		fprintf(stderr, "Could not initialize OpenGL3 render interface.");
+		return false;
+	}
+
+	data->window = target_window;
+	data->gl_context = target_glcontext;
+
+	Rml::SetFileInterface(&data->file_interface);
+	Rml::SetSystemInterface(RmlGui::GetSystemInterface());
+	Rml::SetRenderInterface(RmlGui::GetRenderInterface());
+
+	data->render_interface.SetViewport(winX, winY);
+	data->winX = winX;
+	data->winY = winY;
+
+	Rml::Initialise();
+
+	Rml::LoadFontFace("fonts/FreeSansBold.otf", true);
```

Comment:
Nitpick: Is the code supposed to load two different fonts? It also loads `RmlUi.LoadFontFace("Fonts/FreeMonoBold.ttf", true)` in cont/LuaUI/rml_setup.lua . Please disregard if intended (or maybe add a comment if you want.

Reply:
I forgot that this font was already loaded when I stripped down this file

https://github.com/ChrisFloofyKitsune/Beyond-All-Reason/blob/rmlui-example-widgets/luaui/rml_setup.lua

to make an example rml setup lua stuff for this repo